modules/stochastic_tools/src/actions/LoadCovarianceDataAction.C

```diff
@@ -34,31 +34,90 @@ LoadCovarianceDataAction::act()
   _app.theWarehouse().query().condition<AttribSystem>("SurrogateModel").queryInto(objects);
   for (auto model_ptr : objects)
   {
-    auto * gp = dynamic_cast<GaussianProcess *>(model_ptr);
-    if (gp && model_ptr->isParamValid("filename"))
-      load(*gp);
+    auto * gp_gen = dynamic_cast<GaussianProcessSurrogate *>(model_ptr);
+    if (gp_gen && model_ptr->isParamValid("filename"))
+      load(*gp_gen);
   }
 }
 
 void
-LoadCovarianceDataAction::load(GaussianProcess & model)
+LoadCovarianceDataAction::load(GaussianProcessSurrogate & model)
 {
-  const std::string & covar_type = model.getGPHandler().getCovarType();
-  const std::unordered_map<std::string, Real> & map = model.getGPHandler().getHyperParamMap();
+  // We grab all the necessary information that is needed to reconstruct the
+  // covariance structure for the GP
+  const std::string & covar_type = model.getGP().getCovarType();
+  const std::string & covar_name = model.getGP().getCovarName();
+  const std::map<UserObjectName, std::string> & dep_covar_types =
+      model.getGP().getDependentCovarTypes();
+  const std::vector<UserObjectName> & dep_covar_names = model.getGP().getDependentCovarNames();
+
+  // This is for the covariance on the very top, the lower-level covariances are
+  // all assumed to have num_outputs=1.
+  const unsigned int num_outputs = model.getGP().getCovarNumOutputs();
+  const std::unordered_map<std::string, Real> & map = model.getGP().getHyperParamMap();
   const std::unordered_map<std::string, std::vector<Real>> & vec_map =
-      model.getGPHandler().getHyperParamVectorMap();
-  const UserObjectName & covar_name = model.name() + "_covar_func";
+      model.getGP().getHyperParamVectorMap();
+
+  // We start by creating and loading the lower-level covariances if they need
+  // to be present. Right now we can only load a complex covariance which has
+  // a one-level dependency depth.
+  // TODO: Extend this to arbitrary dependency depths. Maybe we could use a graph.
```

Comment:
Can this consider summation of covariances for a single output GP? Like squared exponential + linear?

Reply:
Yeah but for that you would have to write your new kernel, something like: `SumCovariance` which will depend on as many covariances as you need.